plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/FirKotlinConverter.kt

```diff
@@ -0,0 +1,210 @@
+        requiredTypes: Array<out Class<out UElement>>
+    ): Sequence<UElement> {
+        val ktLightClass = element.toLightClass() ?: return emptySequence()
+        val uClass = AbstractFirKotlinUClass.create(ktLightClass, givenParent)
+        return requiredTypes.accommodate(
+            // Class
+            alternative { uClass },
+            // Object
+            alternative primaryConstructor@{
+                val primaryConstructor = element.primaryConstructor ?: return@primaryConstructor null
+                uClass.methods.asSequence()
+                    .filter { it.sourcePsi == primaryConstructor }
+                    .firstOrNull()
+            }
+        )
+    }
+
+    private fun convertNonLocalProperty(
+        property: KtProperty,
+        givenParent: UElement?,
+        requiredTypes: Array<out Class<out UElement>>
+    ): Sequence<UElement> =
+        requiredTypes.accommodate(*convertToPropertyAlternatives(LightClassUtil.getLightClassPropertyMethods(property), givenParent))
+
+    private fun convertToPropertyAlternatives(
+        methods: LightClassUtil.PropertyAccessorsPsiMethods?,
+        givenParent: UElement?
+    ): Array<UElementAlternative<*>> =
+        if (methods != null)
+            arrayOf(
+                alternative { methods.backingField?.let { FirKotlinUField(it, (it as? KtLightElement<*, *>)?.kotlinOrigin, givenParent) } },
+                alternative { methods.getter?.let { convertDeclaration(it, givenParent, arrayOf(UMethod::class.java)) as? UMethod } },
+                alternative { methods.setter?.let { convertDeclaration(it, givenParent, arrayOf(UMethod::class.java)) as? UMethod } }
+            )
+        else emptyArray()
+
+    internal fun convertPsiElement(
+        element: PsiElement,
+        givenParent: UElement?,
+        requiredTypes: Array<out Class<out UElement>>
+    ): UElement? {
+        fun <P : PsiElement> build(ctor: (P, UElement?) -> UElement): () -> UElement? = {
+            @Suppress("UNCHECKED_CAST")
+            ctor(element as P, givenParent)
+        }
+
+        return with(requiredTypes) {
+            when (element) {
+                is KtImportDirective -> {
+                    el<UImportStatement>(build(::FirKotlinUImportStatement))
+                }
+                else -> null
+            }
+        }
+    }
+}
+
+private inline fun <reified ActualT : UElement> Array<out Class<out UElement>>.el(f: () -> UElement?): UElement? {
+    return if (isAssignableFrom(ActualT::class.java)) f() else null
+}
+
+private inline fun <reified ActualT : UElement> Array<out Class<out UElement>>.expr(f: () -> UExpression?): UExpression? {
+    return if (isAssignableFrom(ActualT::class.java)) f() else null
+}
+
+private fun Array<out Class<out UElement>>.isAssignableFrom(cls: Class<*>) = any { it.isAssignableFrom(cls) }
+
+private fun <U : UElement> Array<out Class<out UElement>>.accommodate(vararg makers: UElementAlternative<out U>): Sequence<UElement> {
```

Comment:
There are already accommodate/alternative functions in FE10 uast. Can we just reuse them?

Reply:
So, the question is, do we want `:plugins:uast-kotlin-fir`, which could replace `:plugins:uast-kotlin`, to depend on `:plugins:uast-kotlin`? That's why I often copy-n-pasted many utils here as-is.

How about introduceing, say, `:plugins:uast-kotlin-common` where shareable utils/declarations can be moved?

Reply:
> How about introduceing, say, :plugins:uast-kotlin-common where shareable utils/declarations can be moved?

let's go this way

Reply:
But we should consider different name than `:plugins:uast-kotlin-common` as it may be confusing with UAST `common` Kotlin code in terms of MPP.

Reply:
Maybe `:plugins:uast-kotlin-base` ?